proposals/nnnn-move-function.md

```diff
@@ -0,0 +1,455 @@
+
+```
+func test() {
+  var x: [Int] = getArray()
+  
+  // x is appended to. After this point, we know that x is unique. We want to
+  // preserve that property.
+  x.append(5)
+  
+  // We create a new variable y so we can write an algorithm where we may
+  // change the value of y (causing a COW copy of the buffer shared with x).
+  var y = x
+  longAlgorithmUsing(&y)
+  // We no longer use y after this point, so move it when we pass it off to
+  // the last use.
+  consumeFinalY(move(y))
+
+  // x will be unique again here.
+  x.append(7)
+}
+```
+
+This addresses both of the motivating issues above: `move` guarantees the
+lifetime of `y` ends at the given point, allowing the compiler to generate
+code to clean up or transfer ownership of `y` without relying on optimization.
+Furthermore, if a future maintainer modifies the code in a way that extends
+the lifetime of `y` past the expected point, then the compiler will raise an
+error. For instance, if we try:
+
+```
+func test() {
+  var x: [Int] = getArray()
+  
+  // x is appended to. After this point, we know that x is unique. We want to
+  // preserve that property.
+  x.append(5)
+  
+  // We create a new variable y so we can write an algorithm where we may
+  // change the value of y (causing a COW copy of the buffer shared with x).
+  var y = x
+  longAlgorithmUsing(&y)
+  // We think we no longer use y after this point...
+  consumeFinalY(move(y))
+
+  // ...and x will be unique again here...
+  x.append(7)
+
+  // ...but this additional use of y snuck in:
+  useYAgain(y)
+}
+```
+
+In this case, we get the following output from the compiler as expected:
+
+```
+test.swift:10:7: error: 'y' used after being moved
+  var y = x
+      ^
+test.swift:13:17: note: move here
+  consumeFinalY(move(y))
+                ^
+test.swift:19:13: note: use here
+  useYAgain(y)
+            ^
+```
+
+Note how the compiler gives all of the information that we need to resolve
+this: it says where the move was and where the later uses that
+cause the problem are, alerting the programmer to what parts of the code are
+trying to extend the lifetime of the value beyond our expected endpoint.
+
+Note that `move` only ends the lifetime of a specific local variable binding.
+It is not tied to the lifetime of the underlying value or any particular
+object instance. If we assign another binding `other` to the same value of `x`,
+we can use that other binding after we end the lifetime of `x`, as in:
+
+```
+func useX(_ x: SomeClassType) -> () {}
+func consumeX(_ x: __owned SomeClassType) -> () {}
+
+func f() -> () {
+  let x = ...
+  useX(x)
+  let other = x   // other is a new binding used to extend the lifetime of x
+  _ = move(x) // x's lifetime ends
+  useX(other)     // other is used here... no problem.
+  consumeX(other) // other is used here... no problem.
+}
+```
+
+In fact, each variable's lifetime is tracked independently, and gets a separate
+diagnostic if used after move. If we try to compile this:
```

Comment:
This is redundant with the paragraphs just above.  I would strike this paragraph, the following example, and the diagnostics after that.